File: tests/functional/Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Skills/WaterfallSkill.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using Microsoft.Bot.Schema;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Skills
{
    public class WaterfallSkill : SkillDefinition
    {
        private enum SkillAction
        {
            Cards,
            Proactive,
            Auth,
            MessageWithAttachment,
            Sso,
            FileUpload,
            Echo,
            Delete,
            Update
        }

        public override IList<string> GetActions()
        {
            return Enum.GetNames(typeof(SkillAction));
        }

        public override Activity CreateBeginActivity(string actionId)
        {
            if (!Enum.TryParse<SkillAction>(actionId, true, out var skillAction))
            {
                throw new InvalidOperationException($"Unable to create begin activity for \"{actionId}\".");
            }

            // We don't support special parameters in these skills so a generic event with the right name
            // will do in this case.
            return new Activity(ActivityTypes.Event) { Name = skillAction.ToString() };
        }
    }
}

File: tests/functional/Bots/Python/Consumers/CodeFirst/WaterfallHostBot/skills/waterfall_skill.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from enum import Enum
from botbuilder.schema import Activity
from skills.skill_definition import SkillDefinition


class WaterfallSkill(SkillDefinition):
    class SkillAction(str, Enum):
        CARDS = "Cards"
        PROACTIVE = "Proactive"
        AUTH = "Auth"
        MESSAGE_WITH_ATTACHMENT = "MessageWithAttachment"
        SSO = "Sso"
        FILE_UPLOAD = "FileUpload"
        ECHO = "Echo"
        DELETE = "Delete"
        UPDATE = "Update"

    def get_actions(self):
        return self.SkillAction

    def create_begin_activity(self, action_id: str):
        if action_id not in self.SkillAction:
            raise Exception(f'Unable to create begin activity for "${action_id}".')

        # We don't support special parameters in these skills so a generic event with the
        # right name will do in this case.
        activity = Activity.create_event_activity()
        activity.name = action_id

        return activity
